### PR TITLE
adding fast audio path when creating buffered queue audio player

### PR DESCRIPTION
**Pull request by ggfan**
please review the fast audio path for buffered queue player creation

### CODE REVIEW

File: native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java [hunk 1]
```diff
@@ -57,9 +59,18 @@ protected void onCreate(Bundle icicle) {
         assetManager = getAssets();
 
         // initialize native audio system
-
         createEngine();
-        createBufferQueueAudioPlayer();
+
+        int sampleRate = 0;
```
Comment (dturner) [inline native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java]: Might be better to default to sensible sample rates and buffer sizes. E.g. 41000 and 1024

File: native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java [hunk 2]
```diff
@@ -57,9 +59,24 @@ protected void onCreate(Bundle icicle) {
...
+        int bufSize = 0;
+        /*
+         * retrieve fast audio path sample rate and buf size; if we have it, we pass to native
+         * side to create a player with fast audio enabled [ fast audio == low latency audio ];
+         * IF we do not have a fast audio path, we pass 0 for sampleRate, which will force native
+         * side to pick up the 8Khz sample rate.
+         */
+        if( android.os.Build.VERSION.SDK_INT > android.os.Build.VERSION_CODES.JELLY_BEAN_MR1) {
```
Comment (claywilkinson) [inline native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java]: can you import Build and not use the full name?

File: native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java [hunk 3]
```diff
@@ -288,7 +305,7 @@ protected void onDestroy()
 
     /** Native methods, implemented in jni folder */
     public static native void createEngine();
-    public static native void createBufferQueueAudioPlayer();
+    public static native void createBufferQueueAudioPlayer(int rate, int size);
```
Comment (claywilkinson) [inline native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java]: can you add a description of size?  size of what?

File: native-audio/app/src/main/jni/native-audio-jni.c [hunk 4]
```diff
@@ -111,6 +114,76 @@ __attribute__((constructor)) static void onDlOpen(void)
...
+}
+
+/*
+ * Only support up-sampling
+ */
+short* createResampledBuf(uint32_t idx, uint32_t srcRate, unsigned *size) {
+    short  *src = NULL;
+    short  *workBuf;
+    int    upSampleRate;
+    int32_t srcSampleCount =0;
```
Comment (claywilkinson) [inline native-audio/app/src/main/jni/native-audio-jni.c]: nit: whitespace

File: native-audio/app/src/main/jni/native-audio-jni.c [hunk 5]
```diff
@@ -243,10 +339,13 @@ void Java_com_example_nativeaudio_NativeAudio_createBufferQueueAudioPlayer(JNIEn
...
-    (void)result;
+    bqPlayerEffectSend = NULL;
+    if( 0 == bqPlayerSampleRate) {
+        result = (*bqPlayerObject)->GetInterface(bqPlayerObject, SL_IID_EFFECTSEND,
+                                                 &bqPlayerEffectSend);
+        assert(SL_RESULT_SUCCESS == result);
+        (void)result;
+    }
 
 #if 0   // mute/solo is not supported for sources that are known to be mono, as this is
```
Comment (claywilkinson) [inline native-audio/app/src/main/jni/native-audio-jni.c]: do you really want #if 0 code in the sample?

File: native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java (hunk 3 above)
Comment (ggfan) [inline native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java]: done: change the parameter  name to reflect the usage:  sampleRate, samplesPerBuffer

File: native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java (hunk 2 above)
Comment (ggfan) [inline native-audio/app/src/main/java/com/example/nativeaudio/NativeAudio.java]: Done

File: native-audio/app/src/main/jni/native-audio-jni.c (hunk 4 above)
Comment (ggfan) [inline native-audio/app/src/main/jni/native-audio-jni.c]: done with new commit

File: native-audio/app/src/main/jni/native-audio-jni.c (hunk 5 above)
Comment (ggfan) [inline native-audio/app/src/main/jni/native-audio-jni.c]: my interpretation of this commented code:  when someone wants to create a multi-channel player,  she/he could easily un-comments the source and uses it right away.

**Comment by claywilkinson** [comment]
lgtm

File: native-audio/app/src/main/jni/native-audio-jni.c (hunk 5 above)
Comment (proppy) [inline native-audio/app/src/main/jni/native-audio-jni.c]: maybe you can change this to be #if MULTI_CHANNEL with a proper define?

**Pull request merged**